src/DependencyInjection/CompilerPass/MiddlewarePass.php

```diff
@@ -116,7 +105,21 @@ private function registerMiddleware(ContainerBuilder $container, array $middlewa
                 $options = array_shift($arguments);
             }
 
-            $options['handler'] = new Reference($clientHandlerStackId);
+            if (!isset($options['handler'])) {
+                $handlerStack = new DefinitionDecorator('csa_guzzle.handler_stack');
+                $handlerStack->setPublic(false);
+
+                $clientHandlerStackId = sprintf('csa_guzzle.handler_stack.%s', $clientId);
+
+                $container->setDefinition($clientHandlerStackId, $handlerStack);
+                $options['handler'] = $handlerStack;
+            }
+
+            $handlerStack = $options['handler'];
+
+            foreach ($clientMiddleware as $middleware) {
+                $handlerStack->addMethodCall('push', [new Reference($middleware['id']), $middleware['alias']]);
```

Comment:
So if its not a HandlerStack you could wrap the custom handler in a HandlerStack first.